ui/de-lib/src/main/java/org/iplantc/de/client/services/impl/DiskResourceServiceFacadeImpl.java

```diff
@@ -391,44 +397,66 @@ public void moveDiskResources(final List<DiskResource> diskResources,
             @Override
             protected DiskResourceMove convertFrom(String result) {
                 DiskResourceMove resourcesMoved = decode(DiskResourceMove.class, result);
+
+                /*
+                 * JDS Result should have a "sources" key
+                 * and a "dest" key.
+                 *
+                 * TODO JDS Verify returned keys to the objects we have already.
+                 */
+
                 // KLUDGE manually set destFolder until services are updated to return full dest info.
                 resourcesMoved.setDestination(destFolder);
-                moveFolders(resourcesMoved);
 
                 return resourcesMoved;
             }
+
+            @Override
+            public void onSuccess(String result) {
+                final DiskResourceMove resourcesMoved = convertFrom(result);
+
+                if (diskResourceUtil.containsFolder(diskResources)) {
+                    onFoldersMoved(sourceFolder, destFolder, resourcesMoved, callback, movedEvent);
+                } else {
+                    eventBus.fireEvent(movedEvent);
+                    callback.onSuccess(resourcesMoved);
+                }
+            }
         });
     }
 
-    private void moveFolders(DiskResourceMove resourcesMoved) {
-        if (resourcesMoved == null || resourcesMoved.getSources() == null) {
-            return;
+    private void onFoldersMoved(Folder sourceFolder,
+                                Folder destFolder,
+                                final DiskResourceMove resourcesMoved,
+                                final AsyncCallback<DiskResourceMove> callback,
+                                final DiskResourcesMovedEvent movedEvent) {
+        if (!diskResourceUtil.isDescendantOfFolder(destFolder, sourceFolder)) {
+            // The source folder is not under the destination, so it needs to be refreshed.
+            refreshFolderOnChildrenMoved(sourceFolder, resourcesMoved, callback, movedEvent);
         }
+        if (!diskResourceUtil.isDescendantOfFolder(sourceFolder, destFolder)) {
+            // The destination is not under the source folder, so it needs to be refreshed.
+            refreshFolderOnChildrenMoved(destFolder, resourcesMoved, callback, movedEvent);
+        }
+    }
```

Comment:
One issue I haven't solved, introduced by this new `onFoldersMoved` method, is now the UI will display 2 "success" notifications when one or more folders are moved from a source directory to a destination that isn't nested under the source, and vice versa. Since this method will call the callback's `onSuccess` method once for each `refreshFolderOnChildrenMoved` call.

For example, if the folder `/a/b/c/d/` is moved into the folder `/a/b/`, then only `/a/b/` will be refreshed and the callback's `onSuccess` method will only be called once after that refresh.
If the folder `/a/b/c/d/` is moved into the folder `/a/e/f/`, then both `/a/b/c/` and `/a/e/f/` will be refreshed and the callback's `onSuccess` will be called after each refresh.